File: ios/Sources/UI/Search/SearchViewModel.swift
import Combine
import CombineSchedulers
import Foundation
import Shared

class SearchViewModel: ObservableObject {
    private let repoRepository: RepoRepository

    private let mainScheduler: AnySchedulerOf<DispatchQueue>

    private var lastQuery: String

    private var cancellables = Set<AnyCancellable>()

    @Published private(set) var searchResultUiModel: SearchResultUiModel

    @Published var query: String

    init(repoRepository: RepoRepository,
         mainScheduler: AnySchedulerOf<DispatchQueue>,
         query: String = "",
         lastQuery: String = "",
         searchResultUiModel: SearchResultUiModel = .init())
    {
        self.repoRepository = repoRepository
        self.mainScheduler = mainScheduler
        self.query = query
        self.lastQuery = lastQuery
        self.searchResultUiModel = searchResultUiModel
    }

    deinit {
        Komol.d("deinit: \(type(of: self))")
    }

    func searchWithNewQuery() {
        guard query != lastQuery else {
            return
        }
        search()
    }

    func search() {
        let query = self.query.lowercased().trimmingCharacters(in: .whitespaces)
        guard !query.isEmpty else {
            lastQuery = query
            searchResultUiModel = SearchResultUiModel()
            return
        }
        guard !searchResultUiModel.inProgress else {
            return
        }

        let mayPaging = query == lastQuery
        let currentSearchResult: PagedRepoSearchResult?
        if mayPaging {
            guard let searchResult = searchResultUiModel.searchResult,
                  searchResult.nextPage != nil
            else {
                return
            }
            currentSearchResult = searchResult
        } else {
            currentSearchResult = nil
        }
        searchResultUiModel = SearchResultUiModel(inProgress: true, searchResult: currentSearchResult)

        let page: Int? = currentSearchResult?.nextPage?.intValue
        Komol.d("query=\(query), page=\(String(describing: page))")
        repoRepository.search(query: query, page: page)
            .receive(on: mainScheduler)
            .sink(with: self) { owner, completion in
                if case let .failure(error) = completion {
                    owner.searchResultUiModel = SearchResultUiModel(
                        searchResult: currentSearchResult,
                        error: error.localizedDescription
                    )
                }
            } receiveValue: { _, newSearchResult in
                defer { self.lastQuery = query }
                if mayPaging {
                    let currentRepos = currentSearchResult?.repos ?? []
                    self.searchResultUiModel = SearchResultUiModel(
                        searchResult: newSearchResult.merge(existingRepos: currentRepos)
                    )
                } else {
                    self.searchResultUiModel = SearchResultUiModel(searchResult: newSearchResult)
                }
            }.store(in: &cancellables)
    }
}

File: android/src/main/java/com/example/android/app/ui/search/SearchViewModel.kt
package com.example.android.app.ui.search

import androidx.annotation.MainThread
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.chrynan.inject.Inject
import com.example.android.app.ui.common.Message
import com.example.shared.data.repository.repo.RepoRepository
import com.example.shared.data.source.local.db.Repo
import com.example.shared.model.GitHubError
import com.example.shared.model.PagedRepoSearchResult
import com.github.droibit.komol.Komol
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

@HiltViewModel
class SearchViewModel(
    val query: MutableStateFlow<String>,
    private var lastQuery: String,
    private val repoRepository: RepoRepository,
    private val searchResultUiModelSink: MutableStateFlow<SearchResultUiModel>
) : ViewModel() {

    val searchResultUiModel: StateFlow<SearchResultUiModel> get() = searchResultUiModelSink

    @Inject
    constructor(
        repoRepository: RepoRepository,
    ) : this(
        query = MutableStateFlow(""),
        lastQuery = "",
        repoRepository = repoRepository,
        searchResultUiModelSink = MutableStateFlow(SearchResultUiModel())
    )

    fun searchWithNewQuery() {
        if (query.value == lastQuery) {
            return
        }
        search()
    }

    @MainThread
    fun search() {
        if (searchResultUiModelSink.value.inProgress) {
            return
        }

        val query = this.query.value.lowercase().filter { !it.isWhitespace() }
        if (query.isEmpty()) {
            lastQuery = query
            emit(SearchResultUiModel())
            return
        }

        val mayPaging = query == lastQuery
        val currentSearchResult: PagedRepoSearchResult? = if (mayPaging) {
            searchResultUiModelSink.value
                .searchResult?.takeIf { it.nextPage != null } ?: return
        } else {
            null
        }
        emit(SearchResultUiModel(inProgress = true, searchResult = currentSearchResult))

        val page = currentSearchResult?.nextPage
        Komol.d("query=$query, page=${page ?: "null"}")

        viewModelScope.launch {
            val result = try {
                val newSearchResult = repoRepository.search(query, page)
                val currentRepos: List<Repo> = if (mayPaging) {
                    currentSearchResult?.repos ?: emptyList()
                } else {
                    emptyList()
                }
                lastQuery = query
                SearchResultUiModel(searchResult = newSearchResult.merge(currentRepos))
            } catch (e: GitHubError) {
                SearchResultUiModel(error = Message(requireNotNull(e.message)))
            }
            emit(result)
        }
    }

    @MainThread
    private fun emit(uiModel: SearchResultUiModel) {
        searchResultUiModelSink.value = uiModel
    }
}